### ISSUE TITLE
Upload images in metadata

### ISSUE THREAD

**Issue by arthur-koenig** [prose/prose#998]
I use Prose with Jekyll and I love the "insert image" tool, it works fine in the markdown editing mode.
However, I am used to writing Front Matter variables to specify paths to images, to be able to use the **same image on different pages**. Is that possible to achieve if the image's path is inserted directly in markdown, as opposed to a variable?
If not, it would be cool to have the exact same "insert image" tool for the metadata fields, which could be specified as an "**element: image**" in the _config.yml file.
What do you think about this?

**Comment by dereklieu** [comment]
If you have an image that you'd like to reuse that's already hosted somewhere, I think the easiest way would be to just write the markdown syntax for images directly into the editor.

`![Alt text](/path/to/img.jpg)`

I believe you can also already do this via metadata, for example if you specify an `image_path` text field that gets inserted into your template.

**Comment by arthur-koenig** [comment]
I should have been more specific, sorry. 
Using the markdown syntax is far from ideal: If I create a new post, I have to use the "insert image" tool twice: once on the post, and once on the page that lists all the posts...
I know I can still avoid that on Prose via metadata, but the picture has to be already uploaded and placed in the right folder. My idea was that one could use Prose to do all that without having to go through github.

**Comment by dereklieu** [comment]
Ah, thanks for the explanation. While I understand the use case much better, I don't think this is something I can use resources to build or support at the moment. Here's why:

1. I don't like the idea of having two identical image uploading workflows, it's possibly confusing and wrongly enforces the idea of "embedding" an image in a document, when really you are just including a metadata image tag.
2. While I understand your workflow is not ideal right now, I think you could find ways to improve it using what's existing. For example, include a metadata text field called `image_url`, upload your image the way you would normally into the body of the post, then copy the parsed url and paste it into the metadata field. 

Thanks for opening this ticket, and please continue to help improve this codebase by raising suggestions and bugs!

**Comment by waatltd** [comment]
I agree with @arthur-koenig that this feature is really important. It would be good to rethink it again. I reviewed many content management interfaces for Jekyll and the biggest disadvantage of Prose.io i found is lack of image upload field, which is one of a key CMS features.

Image display in teasers is one thing (i found already it can be extracted from content using Jekyll hack described there https://blog.omgmog.net/post/extract-the-first-image-from-a-post-in-jekyll/ ) but this is a workaround.

Different part is that some layouts requires to split image from the main content of a post. The best and the most user friendly way to achieve that is via separated meta field.

**Comment by frontcraft** [comment]
I agree with @arthur-koenig that this feature is really important. It would be good to rethink it again. I reviewed many content management interfaces for Jekyll and the biggest disadvantage of Prose.io i found is lack of image upload field, which is one of a key CMS features.

Image display in teasers is one thing (i found already it can be extracted from content using Jekyll hack described there https://blog.omgmog.net/post/extract-the-first-image-from-a-post-in-jekyll/ ) but this is a workaround.

Different part is that some layouts requires to split image from the main content of a post. The best and the most user friendly way to achieve that is via separated meta field.

**Comment by DarkSnow** [comment]
I realise this is a closed issue but this exact issue is stopping me migrating a simple Drupal site to Jekyll.

In these days of responsive design people are moving away from embedding media is WYSIWYG fields in favour of adding that media in a separate field to allow more flexibility in rendering. Am "image" field type in prose.io which allows access to the media browser and stores the path to the image would allow that flexibility in prose.

**Comment by enkota** [comment]
Agree with @DarkSnow. Say a client was going in to update an metadata image. An upload field would make it much easier!

**Comment by pnochisaki** [comment]
I wholeheartedly agree an image upload widget for the yaml front matter would make prose REALLY friggin awesome. I am trying to work prose into a few projects at the moment, but the lack of image insertion to yaml front matter is breaking the deal for most clients/users who will actually be making the edits.  I have them uploading to the body, and pasting the resulting url into the meta data, but they find it cumbersome. Especially people moving away from Drupal/WP who are accustomed to a dedicated image upload "field."  

I agree with @dereklieu that the use case is slightly different when "embedding" and image into a block of copy, vs. specifying a dedicated image in metadata, that could be displayed in multiple ways. I have another concern, however.  Do we really want large image files clogging up our git repos? No, not really.

So, what would be really cool would be developing a cloudinary (or similar) upload widget that uploads the image right into cloudinary, where it can be processed, cropped, hosted etc. and then called directly from the meta data into the place you want it in your templates.

**Comment by pnochisaki** [comment]
OK, here we go... Cloudinary metadata/frontmatter uploader...  Take 1! 

https://www.youtube.com/watch?v=yl8iLkkZqxc

I put this together last night.  And I must say, wow, prose is very easy to work on!  I was a bit concerned that I wouldn't be able to jump in and hit the ground running, especially since I have no experience with backbone.js.  Anyway, that video shows where it's going...

Unfortunately, all my front-end JS is pasted into index.html, because i couldn't figure out how to add it properly in the new meta/cloundinary.js custom view i created. But more on that later.  I am going to reach out to some friends for JS help there and I'll let you all know how it goes, perhaps fork this and get it going... 

But in the meantime, this is pretty much just a general, "Hey what do you think?" I'm curious if this super basic bare-bones UI is looking good to folks?

Basically in _prose.yml you specify all the details at the "field" level
    - field: {api_key: your_api_key, cloud_name: your_cloud_name_here, element: cloudinary,
        help: Select an image here and it will be directly uploaded to Cloudinary,
        label: My Image, upload_preset: your_upload_preset}
      name: my_image

Though one could argue that the api_key, cloud_name, and upload_preset should be in a designated at a higher level in the prose metadata, like this:

prose:
  cloudinary_api_key: your_api_key
  cloudinary_cloud_name: your_cloud_name_here
  cloudinary_upload_preset: your_upload_preset

NOTE: Yes, this method assumes you are OK with "unsigned" uploads in cloudinary, and thus you MUST create a custom upload_preset that specifies this.  the whole idea (for me) here is to be "serverless" and I'm ok with unsigned uploads for the time being.

**Comment by dereklieu** [comment]
@pnochisaki very cool work. I like the direction of the UI and I think it presents a user flow. I'm not super comfortable tying prose.io to a service like cloudinary, but if this functionality worked similar to how prose currently handles image uploads (either from a url or as a commit into the repo) then I'd be happy to work with you on this.

I read your earlier point about clogging up git repos with images and I understand where you're coming from. However, in practice Github allows a very high limit for repo sizes and for the non-technical user that represents a lot of this userbase, it's a perfectly adequate solution, in addition to not requiring anything of that user besides a Github account.

**Comment by pnochisaki** [comment]
OK, all makes sense to me. Yay, let's proceed!  The only thing wrong with my UI pattern in the video above is that it assumes a small thumbnail will be available for instant user feedback that their photo made it to the server.  If we are pushing direct to the repo, we may want to display just an icon, along with the filename, that conveys a successful upload.  I suppose we should consider other files uploads, too, like pdfs.  Having said this, the metadata type seems like it should be "file" or "media".  thoughts on that?

Shall I create a fork, and issue a pull request?  Setting up the views for the new meta type is something I can do quickly to get the ball rolling, but will definitely need help after that.

**Comment by olgageletina** [comment]
Hi there! Great job @pnochisaki! Has this been implemented or is cloudinary the way to go?

**Comment by olgageletina** [comment]
And just a +1 to the comments above – this feature is hey to making Prose a superior tool.
